File: lib/data/repositories/auth_repository.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'auth_repository.g.dart';

@Riverpod(keepAlive: true)
AuthRepository authRepository(Ref ref) =>
    AuthRepository(FirebaseAuth.instance);

class AuthRepository {
  final FirebaseAuth _auth;
  const AuthRepository(this._auth);

  User? get currentUser => _auth.currentUser;

  Future<User> login(String email, String password) async {
    final result =
        await _auth.signInWithEmailAndPassword(email: email, password: password);
    return result.user!;
  }

  Future<User> register(String email, String password) async {
    final result = await _auth.createUserWithEmailAndPassword(
        email: email, password: password);
    return result.user!;
  }

  Future<void> sendPasswordReset(String email) =>
      _auth.sendPasswordResetEmail(email: email);

  Future<void> signOut() => _auth.signOut();
}

File: android_legacy/app/src/main/java/com/example/studytime/data/repository/AuthRepository.kt
package com.example.studytime.data.repository

import com.example.studytime.utils.Resource
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(private val auth: FirebaseAuth) {

    val currentUser: FirebaseUser? get() = auth.currentUser

    suspend fun login(email: String, password: String): Resource<FirebaseUser> = try {
        val result = auth.signInWithEmailAndPassword(email, password).await()
        val user = result.user ?: return Resource.Error("Login succeeded but user is null")
        Resource.Success(user)
    } catch (e: Exception) {
        Resource.Error(e.message ?: "Login failed")
    }

    suspend fun register(email: String, password: String): Resource<FirebaseUser> = try {
        val result = auth.createUserWithEmailAndPassword(email, password).await()
        val user = result.user ?: return Resource.Error("Registration succeeded but user is null")
        Resource.Success(user)
    } catch (e: Exception) {
        Resource.Error(e.message ?: "Registration failed")
    }

    suspend fun sendPasswordReset(email: String): Resource<Unit> = try {
        auth.sendPasswordResetEmail(email).await()
        Resource.Success(Unit)
    } catch (e: Exception) {
        Resource.Error(e.message ?: "Could not send reset email")
    }

    fun signOut() = auth.signOut()
}
